server/src/main/java/org/elasticsearch/transport/ConnectionManager.java

```diff
@@ -166,44 +155,28 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
                                     connectionListener.onNodeDisconnected(node);
                                 }));
                             }
-                            if (conn.isClosed()) {
-                                throw new NodeNotConnectedException(node, "connection concurrently closed");
-                            }
-                            success = true;
-                            listeners = connectingNodes.remove(node);
                         }
-                    } catch (ConnectTransportException e) {
-                        throw e;
-                    } catch (Exception e) {
-                        throw new ConnectTransportException(node, "general node connection failure", e);
                     } finally {
-                        if (success == false) { // close the connection if there is a failure
-                            logger.trace(() -> new ParameterizedMessage("failed to connect to [{}], cleaning dangling connections", node));
-                            IOUtils.closeWhileHandlingException(conn);
-                        } else {
-                            releaseOnce.run();
-                            ActionListener.onResponse(listeners, null);
-                        }
+                        CompletableContext<Void> completableContext = pendingConnections.remove(node);
```

Comment:
Should we call pendingConnections.remove(node, currentAttemptContext) here? This makes sure we only remove it for the right context.
Same goes for the other calls to remove ...

Reply:
I have added a different assertion to ensure that the context returned is always the expected context. The context should not change as a new one cannot be associated with the discovery node key until the existing map entry is removed.